gitlab/resource_gitlab_branch_protection.go

```diff
@@ -34,39 +39,231 @@ func resourceGitlabBranchProtection() *schema.Resource {
 			"merge_access_level": {
 				Type:         schema.TypeString,
 				ValidateFunc: validateValueFunc(acceptedAccessLevels),
-				Required:     true,
+				ExactlyOneOf: schemaMergeExactlyOneOfList,
+				Optional:     true,
 				ForceNew:     true,
 			},
 			"push_access_level": {
 				Type:         schema.TypeString,
 				ValidateFunc: validateValueFunc(acceptedAccessLevels),
-				Required:     true,
+				ExactlyOneOf: schemaPushExactlyOneOfList,
+				Optional:     true,
+				ForceNew:     true,
+			},
+			"unprotect_access_level": {
+				Type:          schema.TypeString,
+				ValidateFunc:  validateValueFunc(acceptedAccessLevels),
+				Default:       accessLevel[gitlab.MaintainerPermissions],
+				ConflictsWith: []string{"allowed_to_unprotect"},
+				Optional:      true,
+				ForceNew:      true,
+			},
+			"allowed_to_merge": {
```

Comment:
This one is a bit complicated.  Maybe things have changed in the intervening months but back when I wrote this, the Terraform SDK didn't support...something (trying to recall, it's been so long...ah here it is: "Using the Elem block to define specific keys for the map is currently not possible.").  After searching around, the work-around that folks were using was a list of max 1 with a custom element that was a map[<type of key>].  

So the maps are type string since "user_id", "group_id", and "access_level" are all strings that denote what kind of values are coming, and the values themselves have types that should match the GitLab API.  The values of "user_id" and "group_id" are set to type int and the "access_level" is set to type string.

In a Terraform config, the resource would thus look like this:

resource "gitlab_branch_protection" "branch_protect" {
	project = gitlab_project.foo.id
	branch = "branch_foo"
	allowed_to_push {
                group_id = [45, 101, 453]
		access_level = ["maintainer"]
	}
	allowed_to_merge {
                user_id = [1, 2, 3]
                access_level = ["maintainer", "developer"]
    }
	allowed_to_unprotect {
		access_level = ["developer"]
	}
}

Reply:
Thanks, so it is about enforcing fields at the schema level. [This](https://github.com/hashicorp/terraform-plugin-sdk/issues/155) might be the thread you were looking at.

My only concern is if GitLab modifies the API in a way that is legal in terms of the raw JSON, but breaks this translation here. For example, they could add a new field to the `ProtectBranchPermissionOptions` object that isn't mutually exclusive with the others.

Would this work? Forgive me if I'm still not understanding the limitation. 😄 
1. Remove `MaxItems: 1`
2. Change "user_id" `Type` to int
3. Change "group_id" `Type` to string
4. Change "access_level" `Type` to string
5. Move your `ExactlyOneOf` check into the inner schema with these attributes.

[This](https://github.com/terraform-providers/terraform-provider-aws/blob/master/aws/resource_aws_cloudfront_distribution.go) is the AWS provider example that was mentioned in that other issue. Linking it as a reference.

Reply:
Yeah that thread is the one I looked at; apparentlymart's first response is very familiar.  Is this what you were thinking?

```
resource "gitlab_branch_protection" "branch_protect" {
	project = gitlab_project.foo.id
	branch = "branch_foo"
	allowed_to_push {
                {
                        group_id = 45
                }
                {
                        group_id = 101
                }
                {
                        group_id = 453
                }
                {
               		access_level = "maintainer"
                }
	}
	allowed_to_merge {
                {
                        user_id = 1
                }
                {
                        user_id = 2
                 }
                 {
                        user_id = 3
                 }
                 {
                        access_level = "developer"
                 }
                 {
                        access_level = "maintainer"
                 }
        }
	allowed_to_unprotect {
		access_level = "developer"
	}
}
```

Basically each of the allowed_to_* attributes can have multiple maps as values, and each map can only specify one value for one permission type.  If you were thinking of something else, please let me know.  This would work but it's very messy...a single list makes far more sense in my opinion...I mean just look at these two examples and tell me which one you'd rather maintain :)

I do see your point regarding the design principles though.  GitLab's API uses a list of single element maps whereas I flipped it around here and made it a map of lists.  I feel like this is a pretty minor change and the improvement in resource clarity justifies the deviation.  Even if GitLab changes the API, we'd still have to update the xanzy/go-gitlab library and that in turn would require changes to this resource.  The mapping between resource attributes to the underlying go-gitlab object is already encapsulated in "constructProtectBranchPermissionOptions".

Reply:
Yeah I see your point. This is what I'm proposing actually:

```terraform
resource "gitlab_branch_protection" "branch_protect" {
  project = "project_foo"
  branch = "branch_foo"
  allowed_to_push {
    group_id = 45
  }
  allowed_to_push {
    group_id = 101
  }
  allowed_to_push {
    access_level = "maintainer"
  }
  allowed_to_push {
    user_id = 1
  }
  allowed_to_push {
    user_id = 2
  }
  allowed_to_unprotect {
    access_level = "developer"
  }
  allowed_to_unprotect {
    user_id = 2
  }
}
```

That's taking what you have, but changing the inner lists to their element types, and removing the `MaxItems: 1` limit from the blocks.

I agree that as a user of this provider, if I had a list of user IDs, I would rather write:

```terraform
locals {
  user_ids = [1, 2, 3]
}

resource "gitlab_branch_protection" "branch_protect" {
  project = "project_foo"
  branch = "branch_foo"
  allowed_to_push {
    user_ids = local.user_ids
  }
}
```

Than:

```terraform
locals {
  user_ids = [1, 2, 3]
}

resource "gitlab_branch_protection" "branch_protect" {
  project = "project_foo"
  branch = "branch_foo"
  dynamic "allowed_to_push" {
    for_each = local.user_ids
    content {
      user_id = allowed_to_push.value
    }
  }
}
```

That said.....

> Even if GitLab changes the API, we'd still have to update the xanzy/go-gitlab library and that in turn would require changes to this resource.

Yes, but they would be backward-compatible changes. My concern is if GitLab added a new field that interacted somehow with `user_id`, for example, there would be no way to add it to the current schema design without making a backward-incompatible change or some kind of messy workaround.

GitLab could have designed their API using the simpler schema you chose, but they didn't. If they want the flexibility of adding new fields, then we should mirror that flexibility in the provider. It's unfortunate when you have to weigh usability and maintainability, and I don't like it either, but in this case I'd still ultimately side with the following the API design.

Reply:
We can see if @nicholasklick or @roidelapluie wants to add any opinions. 😄 It's a tough design call to make.

Reply:
We should use a list of single element maps . Terraform should be as close as the API as possible